File: 2_arrays_part2/1_rotate_matrix.py
#Using extra memory approach
def method(matrix):
    """
    logic - i,j = j,n-i-1
    """

    n = len(matrix)
    res = [[0]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            res[j][n-i-1] = matrix[i][j]
    return res

#Better approach without using extra memory
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        1. Transpose the matrix
        2. Reverse each row
        """
        for i in range(len(matrix)):
            for j in range(i+1, len(matrix[0])):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        for i in range(len(matrix)):
            matrix[i] = matrix[i][::-1]
        return matrix
        

File: 02_arrays_part2/1_rotate_matrix.rs
impl Solution {
    pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
        let n = matrix.len();
        let m = matrix[0].len();

        for i in 0..n {
            for j in i + 1..m {
                let temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

        for i in 0..n {
            matrix[i].reverse();
        }
    }
}
